quarkchain/cluster/tests/test_shard_state.py

```diff
@@ -80,6 +81,72 @@ def test_shard_state_simple(self):
             {self.genesis_token: 2500000000000000000},
         )
 
+    def test_get_total_balance(self):
+        id_list = [Identity.create_random_identity() for _ in range(66)]
+        acc_list = [Address.create_from_identity(i, full_shard_key=0) for i in id_list]
+        random_acc_list = random.sample(range(1, len(acc_list)), len(acc_list) - 1)
+        batch_size = [1, 2, 3, 4, 6, 66]
+        env = get_test_env(
+            genesis_account=acc_list[0], genesis_minor_quarkash=100000000,
+        )
+
+        qkc_token = token_id_encode("QKC")
+
+        state = create_default_shard_state(env=env)
+        # Add a root block to have all the shards initialized
+        root_block = state.root_tip.create_block_to_append().finalize()
+        state.add_root_block(root_block)
+        nonce = 0
+        for i in random_acc_list:
+            tx = create_transfer_transaction(
+                shard_state=state,
+                key=id_list[0].get_key(),
+                from_address=acc_list[0],
+                to_address=acc_list[i],
+                value=100,
+                transfer_token_id=qkc_token,
+                gas_price=0,
+                nonce=nonce,
+            )
+            self.assertTrue(state.add_tx(tx), "the %d tx fails to be added" % i)
+            nonce += 1
+
+        b1 = state.create_block_to_mine(address=acc_list[0])
+        state.finalize_and_add_block(b1)
+
+        self.assertEqual(
+            state.get_token_balance(acc_list[0].recipient, self.genesis_token),
+            100000000 - 6500 + self.get_after_tax_reward(self.shard_coinbase),
+        )
+        self.assertEqual(
+            state.get_token_balance(acc_list[1].recipient, self.genesis_token), 100,
```

Comment:
interesting. I think we may have random failures here (though the probability is small)